File: src/com/sjsu/cs146/labs/lab8/lab8.py
from typing import List

class Solution: 

    def coinChange(self, coins: List[int], amount: int):
        table = [-1] * (amount + 1)
        table[0] = 0
        return self.helper(coins, amount, table)
    
    def helper(self, coins:List[int], amount: int, table: List[int]):        
        if not (table[amount] == -1): return table[amount]
        count = 10000

        for i in range(len(coins)):
            if amount - coins[i] >= 0:
                curr = self.helper(coins, amount - coins[i], table)
                if (not curr == 10000) and curr + 1 < count: count = curr + 1

        table[amount] = count
        return table[amount]

sol = Solution()
coins = [1, 2, 5]
amount = 11
print(sol.coinChange(coins, amount))

    

File: src/com/sjsu/cs146/labs/lab8/lab8.java
package com.sjsu.cs146.labs.lab8;
import java.util.*;

public class lab8 {
    public static void main(String[] args) {
        int[] coins = new int[]{1, 2, 5};
        System.out.println(coinChange(coins, 11));
    }

    public static int coinChange(int[] coins, int amount) {
        int[] table = new int[amount + 1];
        Arrays.fill(table, -1);
        table[0] = 0;
        return helper(coins, amount, table);
    }
    public static int helper(int[] coins, int amount, int[] table) {
        if (table[amount] != -1) return table[amount];

        int count = 1 << 30;

        for (int i = 0; i < coins.length; i++) {
            if (amount - coins[i] >= 0) {
                int curr = helper(coins, amount - coins[i], table);
                if (curr != 1 << 30 && curr + 1 < count) count = curr + 1;
            }
        }
        table[amount] = count;
        return table[amount];
    }
}
